### PR TITLE
feat: add LFM2-VL / LFM2.5-VL vision support

### PR DESCRIPTION
**Pull request by ykhrustalev**
**Problem**
LFM2 worked here as a text-only model. The LFM2-VL models, which also read pictures,
did not work at all. Nothing turned a picture into something the model could read, and
nothing made room for it in the prompt.

**Solution**
- Adds picture handling for LFM2-VL: shrink the picture to a size the model likes, cut it into small squares, and pass those along with a note saying which squares are real picture and how they were laid out.
- Swaps each `<image>` marker in the prompt for the right number of empty slots, so the picture and the text line up. The prompt has to carry exactly one `<image>` per picture, in picture order; any other count is refused up front, the same rule the Hugging Face processor applies, rather than guessed at. One prompt at a time: a one-item list works, longer lists are refused instead of silently losing the text.
- Registers the two new model names, so the loader knows how to run them: one for the picture model, one for its text half.
- Adds a setting for how many squares each picture gets padded to.
- Adds a tutorial and a ready-made config file. Known gaps: a very large picture is shrunk rather than cut into tiles, so it loses some detail, and a picture more than 200 times wider than it is tall (or the reverse) is refused by the resize step.

**Testing**
- New tests check the counting: how many squares a picture turns into, and how many prompt slots it needs.
- A tiny LFM2-VL model is checked in under `test/models/lfm2-vl`, so CI runs the whole pipeline: the picture processor, the hand-off into the decoder, and generation compared against a plain ONNX Runtime replay, with and without a picture and across prefill chunk boundaries. It also pins the refusals above, the fp16 picture-input path, and the pixel values against the Hugging Face processor when `transformers` and `torchvision` are installed (added to the CPU test requirements, except on Windows ARM64 where no `torchvision` wheel exists).
- Ran all six published models on Linux CPU, LFM2-VL 450M, 1.6B, 3B and LFM2.5-VL 450M, 1.6B, 3B, and compared against the original PyTorch version. What we hand the model is the same, except pixel values, which differ by at most 1 out of 255.
- One picture, two pictures, and no picture all give sensible answers. The smallest model writes the same caption as PyTorch, word for word.

**Also in this PR**
- The Windows ARM64 job now force-reinstalls the freshly built wheel. The self-hosted runner kept an older install with the same version number, so pip skipped the new one and the Python tests ran against a stale build.
- `IsLFM2`, the model-type list and the rewind guard are also touched by #2575 (LFM2-MoE); whichever lands second needs a small rebase so `IsLFM2` covers `lfm2`, `lfm2_moe` and `lfm2_vl_text`.

### CODE REVIEW

File: src/models/model_type.h [hunk 1]
```diff
@@ -17,13 +17,13 @@ namespace Generators {
...
+    static constexpr std::array<std::string_view, 32> LLM = {"chatglm", "decoder", "ernie4_5", "gemma", "gemma2", "gemma3_text", "gemma3_vl_text", "gemma4_text", "gpt2", "gptoss", "granite", "granitemoehybrid", "hunyuandensev1", "internlm2", "lfm2", "lfm2_vl_text", "llama", "mistral", "mistral3_text", "nemotron", "olmo", "phi", "phimoe", "phi3", "phi3small", "qwen2", "qwen2_5_vl_text", "qwen3", "qwen3_vl_text", "qwen3_5_moe_text", "qwen3_5_text", "smollm3"};
     return std::find(LLM.begin(), LLM.end(), model_type) != LLM.end();
   }
 
   inline static bool IsVLM(const std::string& model_type) {
     // Vision-language model (VLM)
-    static constexpr std::array<std::string_view, 9> VLM = {"fara", "gemma3", "mistral3", "phi3v", "qwen2_5_vl", "qwen3_vl", "qwen3_5", "qwen3_5_moe", "videochat_flash_qwen"};
+    static constexpr std::array<std::string_view, 10> VLM = {"fara", "gemma3", "lfm2_vl", "mistral3", "phi3v", "qwen2_5_vl", "qwen3_vl", "qwen3_5", "qwen3_5_moe", "videochat_flash_qwen"};
```
[the remark this answers is not in the document]
Comment (ykhrustalev) [reply src/models/model_type.h:26]: Fixed in e3a968a0: the guard at the top of `RewindToLength` now covers `lfm2`, `lfm2_vl_text` and `lfm2_vl`.

File: test/cpp/lfm2_vl_image_geometry_test.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <gtest/gtest.h>
+
+#include <string>
+#include <vector>
+
+#include "models/preprocessing/lfm2_vl_image_processor.h"
```
[the remark this answers is not in the document]
Comment (ykhrustalev) [reply test/cpp/lfm2_vl_image_geometry_test.cpp:10]: Fixed in e3a968a0: moved to `engine_unit_tests`, which links `onnxruntime-genai-obj`.

File: test/python/models/test_lfm2_vl_tokens.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+    images = og.Images.open(image_path)
+    inputs = processor(prompt="<image>describe this", images=images)
+
+    pixel_values = np.array(inputs["pixel_values"])
+    # The processor keys its outputs by the nominal names; they are mapped to the vision graph's
+    # input names (spatial_shapes, pixel_attention_mask) only when handed to the generator.
+    spatial_shapes = np.array(inputs["image_sizes"])
+    pixel_attention_mask = np.array(inputs["image_attention_mask"])
+    num_image_tokens = int(np.array(inputs["num_image_tokens"]).sum())
```
[the remark this answers is not in the document]
Comment (ykhrustalev) [reply test/python/models/test_lfm2_vl_tokens.py]: Fixed in e3a968a0: all four tensors use `as_numpy()`.

File: src/models/preprocessing/lfm2_vl_image_processor.cpp [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    for (int64_t col = 0; col < geometry.patch_cols; ++col) {
+      float* patch = destination + (row * geometry.patch_cols + col) * patch_dim;
+      for (int64_t y = 0; y < encoder_patch_size; ++y) {
+        const int64_t source_row = row * encoder_patch_size + y;
+        for (int64_t x = 0; x < encoder_patch_size; ++x) {
+          const int64_t source_col = col * encoder_patch_size + x;
+          for (int64_t c = 0; c < channels; ++c) {
+            patch[(y * encoder_patch_size + x) * channels + c] =
+                image[c * channel_stride + source_row * padded_width + source_col];
+          }
```
[the remark this answers is not in the document]
Comment (ykhrustalev) [reply src/models/preprocessing/lfm2_vl_image_processor.cpp]: This is the transformers layout. `convert_image_to_patches` permutes `(0, 2, 4, 3, 5, 1)`, so channel is the innermost axis of a patch: `[y][x][channel]`, which is what this writes. Verified bit-identical to the transformers processor, and the end-to-end captions match; the channel-major layout would not.

File: src/python/py/models/builders/base.py [hunk 5]
```diff
@@ -5645,6 +5646,7 @@ def load_weights(self, input_path):
             auto_class_map = {
                 "ForCausalLM": AutoModelForCausalLM,
                 "gemma3_vl_text": Gemma3ForConditionalGeneration,
+                "lfm2_vl": Lfm2VlForConditionalGeneration,
```
[the remark this answers is not in the document]
Comment (ykhrustalev) [reply src/python/py/models/builders/base.py:5769]: The lookup is a substring match (`if k in self.model_type`), so `lfm2_vl_text` hits the `lfm2_vl` entry; both resolve to `Lfm2VlForConditionalGeneration`.

File: src/models/preprocessing/lfm2_vl_image_processor.cpp [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+  }
+
+  // Images the prompt never referenced still have to be consumed by the decoder, otherwise the
+  // vision features and the placeholder positions would not line up. Put them in front of the text.
+  if (next_image < tokens_per_image.size()) {
+    std::string leading;
+    for (size_t i = next_image; i < tokens_per_image.size(); ++i) {
+      leading += BuildLfm2VlImagePlaceholder(tokens_per_image[i]);
+    }
+    expanded.insert(0, leading);
```
Comment (tianleiwu) [inline src/models/preprocessing/lfm2_vl_image_processor.cpp]: [P2] Preserve image order when only some images have placeholders

When there are fewer markers than images, earlier matches already consumed the first images, but this inserts the remaining images' blocks before them without reordering the vision features. With 256×256 and 512×512 images, `processor("Describe <image>.", images=images)` produces placeholder runs `[256, 64]`, while the corresponding features are concatenated as `[64, 256]`. The embedding scatter therefore mixes features from different images within the first block and associates the wrong features with the explicit marker. Please preserve the image order (for example, append unmatched blocks when some markers were consumed), or reject partial references. Add a regression test with one marker and two images having different token counts.

File: src/models/preprocessing/lfm2_vl_image_processor.cpp [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+  CheckResult(OrtxCreateProcessor(processor_.ToBeAssigned(), processor_config.c_str()));
+
+  config.AddMapping(std::string(Config::Defaults::InputIdsName), config.model.embedding.inputs.input_ids);
+  config.AddMapping(std::string(Config::Defaults::PixelValuesName), config.model.vision.inputs.pixel_values);
+  config.AddMapping(std::string(Config::Defaults::ImageAttentionMaskName), config.model.vision.inputs.attention_mask);
+  config.AddMapping(std::string(Config::Defaults::ImageSizesName), config.model.vision.inputs.image_sizes);
+}
+
+std::unique_ptr<NamedTensors> Lfm2VlImageProcessor::Process(const Tokenizer& tokenizer, const Payload& payload) const {
+  const std::string prompt{payload.prompt};
```
Comment (tianleiwu) [inline src/models/preprocessing/lfm2_vl_image_processor.cpp]: [P2] Handle or reject prompt lists instead of discarding their text

The public Python list overload and `MultiModalProcessor::Process(span<...>)` populate `payload.prompts` and leave `payload.prompt` empty, so reading only this field silently drops every supplied prompt. On the checked-in fixture, `processor("Describe this image.")` produces five text tokens, but `processor(["Describe this image."])` produces `input_ids` with shape `[1, 0]`. With images, it retains only the generated image placeholders and loses the instruction. Please consume `payload.prompts` appropriately, or explicitly reject nonempty prompt lists if batching is unsupported, and cover the singleton-list case.

**Review by tianleiwu — CHANGES_REQUESTED** [review]
Two reproducible prompt-processing issues remain: partial image references misalign placeholder blocks with vision features, and list-form prompts silently lose their text. Details and suggested fixes are inline.

Validation at 6bba046e: rebuilt the native library, Python binding, and C++ test target; 12 C++ tests and 77 Python tests passed, with 4 optional Hugging Face parity tests skipped. Both findings were reproduced with the checked-in model fixture and synthetic images. Full published checkpoints and GPU inference were not run locally.

File: src/models/preprocessing/lfm2_vl_image_processor.cpp (hunk 6 above)
Comment (ykhrustalev) [reply src/models/preprocessing/lfm2_vl_image_processor.cpp]: Fixed in 9218337c: the prompt must now hold exactly one `<image>` per image (the `Lfm2VlProcessor.validate_inputs` rule), so the prepend path is gone. Regression tests: `test_lfm2_vl_rejects_image_token_count_mismatch` (one marker with 256px and 512px images, plus the more-markers and no-images cases) and two new C++ cases.

File: src/models/preprocessing/lfm2_vl_image_processor.cpp (hunk 7 above)
Comment (ykhrustalev) [reply src/models/preprocessing/lfm2_vl_image_processor.cpp]: Fixed in 9218337c: a one-entry `payload.prompts` is used as the prompt, longer lists throw "batched prompts are not supported". Covered by `test_lfm2_vl_single_prompt_list_matches_the_string_form` and `test_lfm2_vl_rejects_batched_prompts`.

**Review by tianleiwu — APPROVED** [review]
The two previously blocking prompt-processing issues are fixed at this head: prompt/image cardinality is validated before expansion, and the singleton prompt-list path now preserves its text while rejecting unsupported batches. The current implementation also has direct coverage for patch layout, Hugging Face pixel parity, builder dispatch, vision input validation, convolution-state generation, and rewind rejection.

Local validation: rebuilt the native test target and Python binding; 13 focused C++ tests, 5 builder tests, 52 token tests, and 30 tiny-model runtime/parity tests passed. GitHub currently reports 13 successful checks with Windows CPU ARM64 still in progress. No remaining actionable findings.
